File: plot/src/main/java/net/fwitz/math/plot/renderer/palette/EscapeTimePalette.java
package net.fwitz.math.plot.renderer.palette;

import java.awt.*;
import java.util.ArrayList;
import java.util.List;

public class EscapeTimePalette implements Palette {
    private static final Color[] COLORS;
    static {
        final List<Color> colors = new ArrayList<>(256);
        colors.add(Color.BLACK);
        for (int b = 0x40; b < 0x100; b += 0x10) {
            colors.add(new Color(0x00, 0x00, b));
        }
        colors.add(new Color(0, 0, 0xFF));
        for (int rg = 0; rg < 0x100; rg += 0x10) {
            colors.add(new Color(rg, rg, 0xFF));
        }
        colors.add(Color.WHITE);
        for (int b = 0xF0; b >= 0; b -= 0x10) {
            colors.add(new Color(0xFF, 0xFF, b));
        }
        for (int g = 0xF0; g >= 0; g -= 0x10) {
            colors.add(new Color(0xFF, g, 0x00));
        }
        for (int r = 0xF0, b = 0x04; r > 0; r -= 0x10, b += 0x04) {
            colors.add(new Color(r, 0x00, b));
        }
        COLORS = colors.toArray(new Color[colors.size()]);
    }

    @Override
    public int size() {
        return COLORS.length;
    }

    @Override
    public Color index(int index) {
        return COLORS[index];
    }
}

File: plot/src/main/kotlin/net/fwitz/math/plot/renderer/palette/EscapeTimePalette.kt
package net.fwitz.math.plot.renderer.palette

import java.awt.Color

object EscapeTimePalette : Palette {
    private val COLORS: List<Color> = run {
        val colors = ArrayList<Color>(256)

        colors.add(Color.BLACK)

        var b = 0x40
        while (b < 0x100) {
            colors.add(Color(0x00, 0x00, b))
            b += 0x10
        }

        colors.add(Color(0, 0, 0xFF))

        var rg = 0
        while (rg < 0x100) {
            colors.add(Color(rg, rg, 0xFF))
            rg += 0x10
        }

        colors.add(Color.WHITE)

        b = 0xF0
        while (b >= 0) {
            colors.add(Color(0xFF, 0xFF, b))
            b -= 0x10
        }

        var g = 0xF0
        while (g >= 0) {
            colors.add(Color(0xFF, g, 0x00))
            g -= 0x10
        }

        var r = 0xF0
        b = 0x04

        while (r > 0) {
            colors.add(Color(r, 0x00, b))
            r -= 0x10
            b += 0x04
        }

        colors
    }

    override val size = COLORS.size

    override operator fun get(index: Int) = COLORS[index]
}